### DISCUSSION TITLE
Does wgpu always require 16 byte aligned arrays?

### DISCUSSION THREAD

**Discussion by spencerkohan** [gfx-rs/wgpu#3156] [Q&A] [answered]
I have a shader which is compiled from GLSL to SpirV via spirv-cross and then used in wgpu via Naga.

The type is defined like so in GLSL:

```
struct LayoutRec {
    uint index;
    uint line;
    uint column;
};

layout(set=1, binding=0, std140) uniform LayoutData {
    LayoutRec records[maxCharCount];
} layouts;
```

The resulting layouts for these types look like this, after the module is created from the SPIR-V source:

```
    [5],
    Type {
        name: Some(
            "LayoutRec",
        ),
        inner: Struct {
            members: [
                StructMember {
                    name: Some(
                        "index",
                    ),
                    ty: [4],
                    binding: None,
                    offset: 0,
                },
                StructMember {
                    name: Some(
                        "line",
                    ),
                    ty: [4],
                    binding: None,
                    offset: 4,
                },
                StructMember {
                    name: Some(
                        "column",
                    ),
                    ty: [4],
                    binding: None,
                    offset: 8,
                },
            ],
            span: 12,
        },
    },

...

    [7],
    Type {
        name: None,
        inner: Array {
            base: [5],
            size: Constant(
                [5],
            ),
            stride: 16,
        },
    },
```

At runtime, I get this error when trying to create a pipeline from this module:

```
Shader validation error: 
  ┌─ GlyphProc_vertex:1:1
  │
1 │ 
  │   naga::Type [7]


    Type [7] '' is invalid
    Array stride 16 does not match the expected 12

```

I'm having a bit of trouble narrowing down the source of this error.  Is this something wrong in my GLSL, or is it possibly something wrong in the way I am calling Naga or SPIR-V Cross?  I.e. where does the array stride value get determined to be 16 rather than 12? 

I know I can solve this by adding a padding variable to my `LayoutRec` struct, but I would like to understand if this is actually a requirement to have 16 byte aligned arrays in wgpu.

**Comment by cwfitzgerald** [comment, marked as answer]
I think you can probably call this a bug in Naga. The spirv/glsl frontends should translate this pattern to a valid module. The reason it's a problem is that glsl auto-pads to 16, whereas wgsl requires everything be properly aligned already
